 # Data Consistency Checks

## This script contains the following points:
### 01. Importing libraries
### 02. Importing Data
### 03. Data Wrangling:
- Dropping Columns
- Renaming Columns
### 04. Data Consistency Checks:
- Mixed-Type Data
- Finding Missing Values
- Addressing Missing Values
- Duplicates
### 05. Export Data

## 01. Importing libraries

In [4]:
# Import libraries
import pandas as pd
import os
import numpy as np
import datetime

## 02. Importing Data

In [5]:
# Define path
%run env_1.ipynb
path = path_1

In [6]:
df_pipelines = pd.read_excel(os.path.join(path, '02 Data', 'Original Data', 'leitungen.xlsx'), index_col = False)

In [7]:
df_failures = pd.read_excel(os.path.join(path, '02 Data', 'Original Data', 'schaeden.xlsx'), index_col = False)

In [8]:
df_pipelines.head()

ID_LEITUNG  LEITUNGSART  ZUSTAND  BAUJAHR    LAENGE  INNENDURCHMESSER  \
0     2061749            3        1      NaN  0.177445             600.0   
1     2038858            3        1      NaN  0.143000             600.0   
2     2038836            3        1      NaN  0.143000             600.0   
3     2061410            3        1      NaN  0.185747             600.0   
4     2038844            3        1      NaN  0.143000             600.0   

   WERKSTOFF  BETTUNG  BODENKONSISTENZ  BODENAGGRESSIVITAET  GRUNDWASSER  \
0          4        1                4                    3          1.0   
1          4        1                4                    3          1.0   
2          4        1                4                    3          1.0   
3          4        1                4                    3          1.0   
4          4        1                4                    3          1.0   

   VERKEHRSLAST  STREUSTROM  
0             1           4  
1             1           3  
2             1           3  
3             1           4  
4             1           3

In [9]:
df_failures.head()

LAENGE  SP_AUSSENBESCHICHTUNG      OID   ID_SCHADEN  ID_LEITUNG  \
0     0.1                    NaN  1423998         1196     2067243   
1     0.3                    NaN  1550114         1443     2074816   
2     0.3                    NaN  1854381  83500002/15     2068165   
3     0.3                    NaN  1660080         1433     2062991   
4     0.3                    NaN  1710741         1990     2056069   

   ID_ARMATUR  ID_ZONE GEMELDET_AM BEHOBEN_AM  PERSONALAUFWAND  ...  LECKRATE  \
0           0     60.0  1981-11-14        NaT              NaN  ...         0   
1           0     15.0  1998-11-12        NaT              NaN  ...         0   
2           0     59.0  2006-07-27        NaT              NaN  ...         0   
3           0      6.0  1998-01-13        NaT              NaN  ...         0   
4           0     27.0  2004-08-10        NaT              NaN  ...         0   

   SP_INNENBESCHICHTUNG  SP_SCHADENSSTELLE  SP_BETTUNG  SP_LEITUNGSZUSTAND  \
0                   NaN                  1         NaN                 NaN   
1                   NaN                  1         NaN                 NaN   
2                   NaN                  1         NaN                 NaN   
3                   NaN                  1         NaN                 NaN   
4                   NaN                  1         NaN                 NaN   

   SP_SCHADENSART  SP_SCHADENSURSACHE  SP_ABLAGERUNGEN  WASSERAUSTRITT  \
0               0                 NaN              NaN               0   
1               0                 NaN              NaN               0   
2               0                 NaN              NaN               0   
3               0                 NaN              NaN               0   
4               0                 NaN              NaN               0   

   SP_WERKSTOFF  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN  

[5 rows x 21 columns]

## 03. Data Wrangling:

### 3.1. Dropping Columns

### I. Dataset 'pipelines'

### Observation:
#### We can remove the following columns since they are not essential for achieving the project's objectives:
- LEITUNGSART, ZUSTAND, BETTUNG

In [10]:
df_pipelines_drop = df_pipelines.drop(columns=["LEITUNGSART", "ZUSTAND", "BETTUNG"])

In [11]:
df_pipelines_drop.head(5)

ID_LEITUNG  BAUJAHR    LAENGE  INNENDURCHMESSER  WERKSTOFF  \
0     2061749      NaN  0.177445             600.0          4   
1     2038858      NaN  0.143000             600.0          4   
2     2038836      NaN  0.143000             600.0          4   
3     2061410      NaN  0.185747             600.0          4   
4     2038844      NaN  0.143000             600.0          4   

   BODENKONSISTENZ  BODENAGGRESSIVITAET  GRUNDWASSER  VERKEHRSLAST  STREUSTROM  
0                4                    3          1.0             1           4  
1                4                    3          1.0             1           3  
2                4                    3          1.0             1           3  
3                4                    3          1.0             1           4  
4                4                    3          1.0             1           3

### II. Dataset 'failures'

### Observation:
#### We should retain the following columns for our upcoming analysis: 
- ID_SCHADEN,ID_LEITUNG, GEMELDET_AM, SP_SCHADENSURSACHE

In [12]:
# Filter columns
columns_to_keep = ["ID_SCHADEN", "ID_LEITUNG", "GEMELDET_AM", "SP_SCHADENSURSACHE"]

In [13]:
df_failures_filtered = df_failures[columns_to_keep]

In [14]:
df_failures_filtered.head(5)

ID_SCHADEN  ID_LEITUNG GEMELDET_AM  SP_SCHADENSURSACHE
0         1196     2067243  1981-11-14                 NaN
1         1443     2074816  1998-11-12                 NaN
2  83500002/15     2068165  2006-07-27                 NaN
3         1433     2062991  1998-01-13                 NaN
4         1990     2056069  2004-08-10                 NaN

### 3.2. Renaming Columns

### I. Dataset 'pipelines'

### Observation:
#### We could rename the following columns because they are in German:
##### 'ID_LEITUNG' to 'id_pipeline'
##### 'BAUJAHR' to 'year_installation'
##### 'INNENDURCHMESSER' to 'diameter'
##### 'LAENGE' to 'length'
##### 'WERKSTOFF' to 'material'
##### 'BODENKONSISTENZ'	to 'soil_consistency'
##### 'BODENAGGRESSIVITAET' to 'soil_aggressivity'
##### 'GRUNDWASSER' to 'groundwater' 
##### 'VERKEHRSLAST' to 'traffic_load' 
##### 'STREUSTROM' to 'stray_electricity'

In [15]:
df_pipelines_drop.rename(columns = {'ID_LEITUNG' : 'id_pipeline', 'LAENGE' : 'length', 'BAUJAHR' : 'year_installation', 'INNENDURCHMESSER' : 'diameter', 'WERKSTOFF' : 'material', 'BODENKONSISTENZ' : 'soil_consistency', 'BODENAGGRESSIVITAET' : 'soil_aggressivity', 'GRUNDWASSER' : 'groundwater', 'VERKEHRSLAST' : 'traffic_load', 'STREUSTROM' : 'stray_electricity' }, inplace = True)

In [16]:
df_pipelines_drop.head(5)

id_pipeline  year_installation    length  diameter  material  \
0      2061749                NaN  0.177445     600.0         4   
1      2038858                NaN  0.143000     600.0         4   
2      2038836                NaN  0.143000     600.0         4   
3      2061410                NaN  0.185747     600.0         4   
4      2038844                NaN  0.143000     600.0         4   

   soil_consistency  soil_aggressivity  groundwater  traffic_load  \
0                 4                  3          1.0             1   
1                 4                  3          1.0             1   
2                 4                  3          1.0             1   
3                 4                  3          1.0             1   
4                 4                  3          1.0             1   

   stray_electricity  
0                  4  
1                  3  
2                  3  
3                  4  
4                  3

### II. Dataset 'failures'

### Observation:
#### We could rename the following columns because they are in German:
##### 'ID_LEITUNG' to 'id_pipeline'
##### 'ID_SCHADEN' to 'id_failure'
##### 'GEMELDET_AM' to 'failure_reported_on'
##### 'SP_SCHADENSURSACHE' to 'cause_of_damage' 

In [17]:
df_failures_filtered = df_failures_filtered.copy()
df_failures_filtered.rename(columns={'ID_SCHADEN': 'id_failure', 'ID_LEITUNG': 'id_pipeline', 'GEMELDET_AM': 'failure_reported_on', 'SP_SCHADENSURSACHE': 'cause_of_damage'}, inplace=True)

In [18]:
df_failures_filtered.head(5)

id_failure  id_pipeline failure_reported_on  cause_of_damage
0         1196      2067243          1981-11-14              NaN
1         1443      2074816          1998-11-12              NaN
2  83500002/15      2068165          2006-07-27              NaN
3         1433      2062991          1998-01-13              NaN
4         1990      2056069          2004-08-10              NaN

## 04. Data Consistency Checks:

### I. Dataset 'pipelines'

In [19]:
df_pipelines_drop.describe()

id_pipeline  year_installation        length      diameter  \
count  8.356100e+04       73583.000000  83561.000000  78809.000000   
mean   1.731185e+06        1983.113273      9.821036     90.718649   
std    3.335894e+05          19.478042     25.452834     67.563176   
min    4.253450e+05        1875.000000      0.000000     25.000000   
25%    1.447842e+06        1972.000000      0.578237     40.000000   
50%    1.840076e+06        1984.000000      4.327451     65.000000   
75%    2.055380e+06        1999.000000     11.330000    100.000000   
max    2.077833e+06        2013.000000   1592.368000    600.000000   

           material  soil_consistency  soil_aggressivity   groundwater  \
count  83561.000000      83561.000000       83561.000000  83251.000000   
mean       4.217266          1.794103           2.158064      1.029381   
std        2.247115          1.731223           1.343567      0.348797   
min        0.000000          0.000000           0.000000      0.000000   
25%        3.000000          0.000000           1.000000      1.000000   
50%        4.000000          1.000000           3.000000      1.000000   
75%        5.000000          3.000000           3.000000      1.000000   
max       10.000000          5.000000           4.000000      3.000000   

       traffic_load  stray_electricity  
count  83561.000000       83561.000000  
mean       1.603427           3.397697  
std        0.802912           1.038602  
min        0.000000           1.000000  
25%        1.000000           3.000000  
50%        1.000000           4.000000  
75%        2.000000           4.000000  
max        5.000000           4.000000

#### Observation of the frequency table: 
#### When certain variables are coded with a value of 0, it signifies that the input is "Unknown." We observe this code for the variables: material, soil consistency, soil aggressivity, groundwater, and traffic load, where the minimum values are 0. This indicates the presence of entries with "Unknown" information.
#### Of particular note is the variable "soil consistency," where as much as 25% of the entries have a value of 0. This underscores the need for careful consideration when dealing with this variable, as a substantial portion of the data is categorized as "Unknown" for soil consistency.

### Mixed-Type Data

In [20]:
# Check for mixed types
for col in df_pipelines_drop.columns:
    weird = (df_pipelines_drop[[col]].applymap(type) != df_pipelines_drop[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_pipelines_drop[weird]) > 0:
        print(col)

#### Observation: There are no mixed data type in the dataframe.

### Finding Missing Values

In [21]:
df_pipelines_drop.isnull().sum()

id_pipeline             0
year_installation    9978
length                  0
diameter             4752
material                0
soil_consistency        0
soil_aggressivity       0
groundwater           310
traffic_load            0
stray_electricity       0
dtype: int64

In [22]:
df_pipelines_drop.shape

(83561, 10)

In [23]:
# Number of zeros and missing values (NaN) in each column
count_zeros_and_missing = (df_pipelines_drop == 0).sum() + df_pipelines_drop.isnull().sum()

In [24]:
count_zeros_and_missing

id_pipeline              0
year_installation     9978
length                  21
diameter              4752
material              5492
soil_consistency     23402
soil_aggressivity     6320
groundwater           2311
traffic_load          2001
stray_electricity        0
dtype: int64

#### Observation: There are 9978 missing values for the age of the pipelines; 21 unknown values for the length of the pipelines; 4752 missing values for the diameter of the pipelines; 5492 unknown values for the pipes material, 23402 unknown values for the soil_consistency, 6320 unknown  for the soil_aggressivity, 2001 unknown and 310 missing for the groundwater and 2001 unknown for the traffic_load.

In [25]:
# List of  rows where "material" "soil consistency" "soil aggressivity" "groundwater" and "traffic load" are all equal to 0, and 'year_installation', 'diameter', and 'groundwater' are NaN
df_zero_nan = df_pipelines_drop[
    (df_pipelines_drop['material'] == 0) &
    (df_pipelines_drop['soil_consistency'] == 0) &
    (df_pipelines_drop['soil_aggressivity'] == 0) &
    (df_pipelines_drop['groundwater'] == 0) &
    (df_pipelines_drop['traffic_load'] == 0) &
    (df_pipelines_drop['year_installation'].isnull()) &
    (df_pipelines_drop['diameter'].isnull()) &
    (df_pipelines_drop['groundwater'].isnull())
]

In [26]:
df_zero_nan

Empty DataFrame
Columns: [id_pipeline, year_installation, length, diameter, material, soil_consistency, soil_aggressivity, groundwater, traffic_load, stray_electricity]
Index: []

#### Observation: There aren't rows where "material," "soil consistency," "soil aggressivity," "groundwater," and "traffic load" are all equal to 0, and 'year_installation', 'diameter', and 'groundwater' are NaN.

In [27]:
# List the rows where all three variables "year_installation" "diameter" and "groundwater" are missing
df_pipe_nan = df_pipelines_drop[df_pipelines_drop[['year_installation', 'diameter', 'groundwater']].isnull().all(axis=1)]

In [28]:
df_pipe_nan

id_pipeline  year_installation  length  diameter  material  \
78929      1870505                NaN   0.313       NaN         0   
79988      1849328                NaN  46.645       NaN         0   
80408      1881762                NaN   0.230       NaN         0   
80718      1870457                NaN   0.563       NaN         0   
81740      1870524                NaN  38.502       NaN         0   
82198      1973010                NaN  10.406       NaN         0   

       soil_consistency  soil_aggressivity  groundwater  traffic_load  \
78929                 0                  0          NaN             1   
79988                 0                  0          NaN             1   
80408                 0                  0          NaN             1   
80718                 0                  0          NaN             1   
81740                 0                  0          NaN             1   
82198                 0                  0          NaN             1   

       stray_electricity  
78929                  4  
79988                  4  
80408                  4  
80718                  4  
81740                  4  
82198                  4

#### Observation: There are 6 rows where almost every varable is missing (NaN or code 0), so i will need to delete those rows.

In [29]:
# List of  rows where "material" "soil consistency" "soil aggressivity" "groundwater" and "traffic load" are all equal to 0
df_zero_values = df_pipelines_drop[
    (df_pipelines_drop['material'] == 0) &
    (df_pipelines_drop['soil_consistency'] == 0) &
    (df_pipelines_drop['soil_aggressivity'] == 0) &
    (df_pipelines_drop['groundwater'] == 0) &
    (df_pipelines_drop['traffic_load'] == 0)
]

In [30]:
df_zero_values

id_pipeline  year_installation     length  diameter  material  \
78854      1769929             2005.0   5.810000       NaN         0   
78855      1774476                NaN   4.587000       NaN         0   
78856      1774576                NaN   7.771000       NaN         0   
78857      1774997                NaN  10.103000       NaN         0   
78858      1778002                NaN   8.138000       NaN         0   
...            ...                ...        ...       ...       ...   
83539      1869835                NaN   4.131000       NaN         0   
83541      1870027                NaN   1.563000       NaN         0   
83543      1879483                NaN  45.075000       NaN         0   
83547      2021528                NaN   2.196000       NaN         0   
83553      2059277                NaN   2.882684       NaN         0   

       soil_consistency  soil_aggressivity  groundwater  traffic_load  \
78854                 0                  0          0.0             0   
78855                 0                  0          0.0             0   
78856                 0                  0          0.0             0   
78857                 0                  0          0.0             0   
78858                 0                  0          0.0             0   
...                 ...                ...          ...           ...   
83539                 0                  0          0.0             0   
83541                 0                  0          0.0             0   
83543                 0                  0          0.0             0   
83547                 0                  0          0.0             0   
83553                 0                  0          0.0             0   

       stray_electricity  
78854                  4  
78855                  4  
78856                  4  
78857                  3  
78858                  3  
...                  ...  
83539                  4  
83541                  4  
83543                  3  
83547                  2  
83553                  4  

[2001 rows x 10 columns]

#### Observation: There are 2001 rows where almost every varable is missing (NaN or code 0), so i need to delete those rows.

### Note: To achievthe specific objectives of the research project, we need to have values in the raw for at least one of the structural factors(pipe age, pipe material, pipe diameter). 

In [31]:
# List the raws where "year_installation" "diameter" are missing and "material" is zero
df_structural_zero = df_pipelines_drop[
    (df_pipelines_drop['material'] == 0) &
    (df_pipelines_drop['year_installation'].isnull()) &
    (df_pipelines_drop['diameter'].isnull())
]

In [32]:
df_structural_zero

id_pipeline  year_installation      length  diameter  material  \
78809      1083128                NaN    3.475000       NaN         0   
78810      1085626                NaN    1.998000       NaN         0   
78811      1085680                NaN    6.225000       NaN         0   
78812      1090857                NaN    9.561000       NaN         0   
78813      1091082                NaN    5.378000       NaN         0   
...            ...                ...         ...       ...       ...   
83554      2059646                NaN    7.891595       NaN         0   
83555      2065346                NaN    0.171406       NaN         0   
83556      2065405                NaN    5.823676       NaN         0   
83559      2075640                NaN  167.889912       NaN         0   
83560      2077634                NaN    1.420479       NaN         0   

       soil_consistency  soil_aggressivity  groundwater  traffic_load  \
78809                 1                  1          1.0             2   
78810                 1                  1          1.0             3   
78811                 1                  1          1.0             3   
78812                 1                  1          1.0             2   
78813                 1                  1          1.0             2   
...                 ...                ...          ...           ...   
83554                 0                  0          1.0             1   
83555                 3                  1          1.0             1   
83556                 3                  3          1.0             1   
83559                 5                  3          1.0             1   
83560                 3                  1          1.0             1   

       stray_electricity  
78809                  4  
78810                  4  
78811                  4  
78812                  3  
78813                  4  
...                  ...  
83554                  4  
83555                  4  
83556                  4  
83559                  4  
83560                  4  

[4507 rows x 10 columns]

#### Observation: There are 4507 rows where all the structural factors(pipe age, pipe material, pipe diameter) are missing, so i need to delete those rows.

### Addressing Missing Values

In [33]:
# Create a new DataFrame by excluding the rows that match the criteria in df_zero_values
df_non_zero_values = df_pipelines_drop[~
    ((df_pipelines_drop['material'] == 0) &
    (df_pipelines_drop['soil_consistency'] == 0) &
    (df_pipelines_drop['soil_aggressivity'] == 0) &
    (df_pipelines_drop['groundwater'] == 0) &
    (df_pipelines_drop['traffic_load'] == 0))
]

In [34]:
df_non_zero_values

id_pipeline  year_installation      length  diameter  material  \
0          2061749                NaN    0.177445     600.0         4   
1          2038858                NaN    0.143000     600.0         4   
2          2038836                NaN    0.143000     600.0         4   
3          2061410                NaN    0.185747     600.0         4   
4          2038844                NaN    0.143000     600.0         4   
...            ...                ...         ...       ...       ...   
83556      2065405                NaN    5.823676       NaN         0   
83557      2073703             2002.0   22.550198       NaN         0   
83558      2073704             2002.0    1.223609       NaN         0   
83559      2075640                NaN  167.889912       NaN         0   
83560      2077634                NaN    1.420479       NaN         0   

       soil_consistency  soil_aggressivity  groundwater  traffic_load  \
0                     4                  3          1.0             1   
1                     4                  3          1.0             1   
2                     4                  3          1.0             1   
3                     4                  3          1.0             1   
4                     4                  3          1.0             1   
...                 ...                ...          ...           ...   
83556                 3                  3          1.0             1   
83557                 0                  4          2.0             1   
83558                 0                  4          2.0             1   
83559                 5                  3          1.0             1   
83560                 3                  1          1.0             1   

       stray_electricity  
0                      4  
1                      3  
2                      3  
3                      4  
4                      3  
...                  ...  
83556                  4  
83557                  4  
83558                  4  
83559                  4  
83560                  4  

[81560 rows x 10 columns]

#### Observation: In the new DataFrame, there are 81.560 rows, which indicates that precisely 2.001 rows were removed from the original DataFrame.

In [35]:
# Create a new DataFrame, df_filtered, that includes only the rows from df_non_zero_values that do not have a match in df_pipe_nan
merged = df_non_zero_values.merge(df_pipe_nan, how='left', indicator=True)
df_filtered = merged[merged['_merge'] == 'left_only'].drop(columns=['_merge'])

In [36]:
df_filtered

id_pipeline  year_installation      length  diameter  material  \
0          2061749                NaN    0.177445     600.0         4   
1          2038858                NaN    0.143000     600.0         4   
2          2038836                NaN    0.143000     600.0         4   
3          2061410                NaN    0.185747     600.0         4   
4          2038844                NaN    0.143000     600.0         4   
...            ...                ...         ...       ...       ...   
81555      2065405                NaN    5.823676       NaN         0   
81556      2073703             2002.0   22.550198       NaN         0   
81557      2073704             2002.0    1.223609       NaN         0   
81558      2075640                NaN  167.889912       NaN         0   
81559      2077634                NaN    1.420479       NaN         0   

       soil_consistency  soil_aggressivity  groundwater  traffic_load  \
0                     4                  3          1.0             1   
1                     4                  3          1.0             1   
2                     4                  3          1.0             1   
3                     4                  3          1.0             1   
4                     4                  3          1.0             1   
...                 ...                ...          ...           ...   
81555                 3                  3          1.0             1   
81556                 0                  4          2.0             1   
81557                 0                  4          2.0             1   
81558                 5                  3          1.0             1   
81559                 3                  1          1.0             1   

       stray_electricity  
0                      4  
1                      3  
2                      3  
3                      4  
4                      3  
...                  ...  
81555                  4  
81556                  4  
81557                  4  
81558                  4  
81559                  4  

[81554 rows x 10 columns]

#### Observation: Six additional rows were removed from the original dataset, which is the same number as the columns in the df_pipe_nan DataFrame. 

In [37]:
# Create new DataFrame, df_pipes_clean, that contains only the rows from df_filtered that do not have a match in df_structural_zero
merged = df_filtered.merge(df_structural_zero, how='left', indicator=True)
df_pipes_clean = merged[merged['_merge'] == 'left_only'].drop(columns=['_merge'])

In [38]:
df_pipes_clean

id_pipeline  year_installation     length  diameter  material  \
0          2061749                NaN   0.177445     600.0         4   
1          2038858                NaN   0.143000     600.0         4   
2          2038836                NaN   0.143000     600.0         4   
3          2061410                NaN   0.185747     600.0         4   
4          2038844                NaN   0.143000     600.0         4   
...            ...                ...        ...       ...       ...   
81507      1802960             1975.0   8.661000       NaN         0   
81513      1846424             2008.0   0.350000       NaN         0   
81537      1869916             2003.0  57.601000       NaN         0   
81550      2073703             2002.0  22.550198       NaN         0   
81551      2073704             2002.0   1.223609       NaN         0   

       soil_consistency  soil_aggressivity  groundwater  traffic_load  \
0                     4                  3          1.0             1   
1                     4                  3          1.0             1   
2                     4                  3          1.0             1   
3                     4                  3          1.0             1   
4                     4                  3          1.0             1   
...                 ...                ...          ...           ...   
81507                 1                  1          1.0             1   
81513                 0                  0          1.0             1   
81537                 1                  1          1.0             1   
81550                 0                  4          2.0             1   
81551                 0                  4          2.0             1   

       stray_electricity  
0                      4  
1                      3  
2                      3  
3                      4  
4                      3  
...                  ...  
81507                  4  
81513                  4  
81537                  4  
81550                  4  
81551                  4  

[78951 rows x 10 columns]

#### Observation: Additional 2603 rows were removed from the original dataset. The final dataframe contains 78951 rows × 9 columns, where in total 4610 rows were deleted.

### Duplicates

In [39]:
df_pipelines_drop_dups = df_pipes_clean[df_pipes_clean.duplicated()]

In [40]:
df_pipelines_drop_dups

Empty DataFrame
Columns: [id_pipeline, year_installation, length, diameter, material, soil_consistency, soil_aggressivity, groundwater, traffic_load, stray_electricity]
Index: []

#### Observation: No duplicates found.

### II. Dataset 'failures'

In [41]:
df_failures_filtered.describe()

id_pipeline  cause_of_damage
count  2.347000e+03      1435.000000
mean   1.839234e+06         2.747735
std    2.923008e+05         0.883239
min    1.010389e+06         0.000000
25%    1.739365e+06         2.000000
50%    1.959787e+06         3.000000
75%    2.062426e+06         3.000000
max    2.077801e+06         7.000000

#### Observation of the frequency table: For the variable  cause_of_damage  the minimum value is 0. This indicates the presence of entries with "Unknown" information.

### Mixed-Type Data

In [42]:
# Check for mixed types
for col in df_failures_filtered.columns:
    weird = (df_failures_filtered[[col]].applymap(type) != df_failures_filtered[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_failures_filtered[weird]) > 0:
        print(col)

id_failure
failure_reported_on


#### Observation: The columns 'id_failure' and 'reported_on' contain mixed data type.

In [43]:
df_failures_filtered['id_failure'].dtype

dtype('O')

#### Observation: The variable is an ID and the dtype('O') indicates that the data type of the particular column is an object (usually a string). I don't need to worry about mixed data types in this case.

In [44]:
df_failures_filtered['failure_reported_on'].dtype

dtype('<M8[ns]')

#### Observation: The dtype('<M8[ns]') indicates that the column contains datetime values, where [ns] represents nanoseconds. I will  change the data type of a datetime column to include only the date part and not the time.

In [45]:
# Change the data type of a datetime column to include only the date part and not the time
df_failures_filtered['failure_reported_on'] = pd.to_datetime(df_failures_filtered['failure_reported_on']).dt.date

In [46]:
df_failures_filtered['failure_reported_on'].dtype

dtype('O')

### Duplicates

In [47]:
# Create a Dataframe with the duplicated rows
df_failures_filtered_dups = df_failures_filtered[df_failures_filtered.duplicated()]

In [48]:
df_failures_filtered_dups.shape

(20, 4)

#### Observation: There are 20 duplicated rows.

In [49]:
# Create a Dataframe that excludes duplicated rows
df_failures_clear = df_failures_filtered.drop_duplicates()

In [50]:
df_failures_clear.shape

(2327, 4)

In [51]:
df_failures_filtered.shape

(2347, 4)

### Finding Missing Values

In [52]:
df_failures_clear.isnull().sum()

id_failure               0
id_pipeline              0
failure_reported_on      4
cause_of_damage        910
dtype: int64

In [53]:
df_failures_clear.shape

(2327, 4)

In [54]:
# Number of zeros and missing values (NaN) in each column
failures_count_zeros_and_missing = (df_failures_clear == 0).sum() + df_failures_clear.isnull().sum()

In [55]:
failures_count_zeros_and_missing

id_failure               0
id_pipeline              0
failure_reported_on      4
cause_of_damage        941
dtype: int64

#### Observation: In the dataset, there are 910 missing values and 31 entries labeled as "unknown" for the 'cause_of_damage' attribute, and 4 missing values for the reported failure date. In total, these missing and unknown values account for approximately 40% of the 'cause_of_damage' attribute. Given the significant number of missing/unknown values in 'cause_of_damage,' I chose not to remove these rows. Removing them would greatly reduce the dataset for failure analysis, potentially making it insufficient for meaningful analysis.

### Addressing Missing Values

### No need of any treatment for the missing values.

### Final check of  data before exporting it

In [56]:
df_pipes_clean['year_installation'].dtype

dtype('float64')

In [57]:
# Define the target year
target_year = 2013

# Calculate pipe age in the specified year
df_pipes_clean['pipe_age_2013'] = target_year - df_pipes_clean['year_installation']

# Display the DataFrame
print(df_pipes_clean)

       id_pipeline  year_installation     length  diameter  material  \
0          2061749                NaN   0.177445     600.0         4   
1          2038858                NaN   0.143000     600.0         4   
2          2038836                NaN   0.143000     600.0         4   
3          2061410                NaN   0.185747     600.0         4   
4          2038844                NaN   0.143000     600.0         4   
...            ...                ...        ...       ...       ...   
81507      1802960             1975.0   8.661000       NaN         0   
81513      1846424             2008.0   0.350000       NaN         0   
81537      1869916             2003.0  57.601000       NaN         0   
81550      2073703             2002.0  22.550198       NaN         0   
81551      2073704             2002.0   1.223609       NaN         0   

       soil_consistency  soil_aggressivity  groundwater  traffic_load  \
0                     4                  3          1.0       

#### Rearrange the order of columns 

In [58]:
# Get a list of column names
column_names = df_pipes_clean.columns.tolist()

In [59]:
column_names

['id_pipeline',
 'year_installation',
 'length',
 'diameter',
 'material',
 'soil_consistency',
 'soil_aggressivity',
 'groundwater',
 'traffic_load',
 'stray_electricity',
 'pipe_age_2013']

In [60]:
# Remove the column from its current position
column_names.remove('pipe_age_2013')

# Insert the column at the desired position (second position in this case)
column_names.insert(2, 'pipe_age_2013')

# Reorder the DataFrame with the new column order
df_pipes_clean = df_pipes_clean[column_names]

In [61]:
df_pipes_clean.head(5)

id_pipeline  year_installation  pipe_age_2013    length  diameter  \
0      2061749                NaN            NaN  0.177445     600.0   
1      2038858                NaN            NaN  0.143000     600.0   
2      2038836                NaN            NaN  0.143000     600.0   
3      2061410                NaN            NaN  0.185747     600.0   
4      2038844                NaN            NaN  0.143000     600.0   

   material  soil_consistency  soil_aggressivity  groundwater  traffic_load  \
0         4                 4                  3          1.0             1   
1         4                 4                  3          1.0             1   
2         4                 4                  3          1.0             1   
3         4                 4                  3          1.0             1   
4         4                 4                  3          1.0             1   

   stray_electricity  
0                  4  
1                  3  
2                  3  
3                  4  
4                  3

In [62]:
df_failures_clear.head(5)

id_failure  id_pipeline failure_reported_on  cause_of_damage
0         1196      2067243          1981-11-14              NaN
1         1443      2074816          1998-11-12              NaN
2  83500002/15      2068165          2006-07-27              NaN
3         1433      2062991          1998-01-13              NaN
4         1990      2056069          2004-08-10              NaN

## 05. Export Data

### I. Dataset 'pipelines'

In [63]:
df_pipes_clean.to_excel(os.path.join(path, '02 Data','Prepared Data', 'pipelines_clean.xlsx'))

### II. Dataset 'failures'

In [64]:
df_failures_clear.to_excel(os.path.join(path, '02 Data','Prepared Data', 'failures_clean.xlsx'))